### PR TITLE
allow optional disabling of interning within remote-write QueueManager

### PR DESCRIPTION
**Pull request by cstyan**
****Note:** This change allows for the optional disabling of interning within the QueueManager. This is not related to the interning of labels string that must happen as part of a v2 proto WriteRequest, but only to Prometheus' own code implementation of remote write which in the past needed this additional interning code path to reduce memory allocated when receiving labels data from the WAL Watcher routine. The changes here are really only relevant to downstream importers of Prometheus' Remote Write implementation, such as the Loki Ruler.**

Supersedes #14748 

I was curious as to whether the new unique package in go 1.23 could help us anywhere. I knew the Grafana agent/alloy team had investigated having a single interning map across all configured remote writes, but that also increase the complexity of managing which interned strings should still be present, and which should be removed because of compaction/truncation of the WAL. 

However, after spending some time trying various changes to the interning code as well as the map within `queue_manager.go` that stores the labels, the conclusion I'm coming to is that the interning is no longer necessary for remote write. Because we enqueue samples with their entire labelset, we no longer benefit from the string interning because we are not recreating the labelset when building the write request.

We could potentially benefit from a shared cache of the labelsets if multiple remote writes are being used, but only if they have the same relabel rules and external labels. If they didn't the cached labelset would be incorrect for one or the other, or we'd overwrite the entry in the map. Either way,  I think it's optimizing an edge case that is rarely used and the improvements from just removing the interning are already very good.

Additionally, these changes could have knock on effects for projects that reuse the remote write code. For example, loki's ruler component runs a remote write per tenant who has rules configured.

```
                                                   │     base      │                no-intern                 │               split-intern               │
                                                   │    sec/op     │    sec/op     vs base                    │    sec/op     vs base                    │
SampleSend-32                                         3.636m ±  5%
StoreSeries/plain-32                                  901.4µ ± 10%   203.0µ ± ∞ ¹  -77.48% (p=0.001 n=10+5)     204.0µ ± ∞ ¹  -77.37% (p=0.001 n=10+5)
StoreSeries/externalLabels-32                        1777.1µ ±  5%   823.2µ ± ∞ ¹  -53.68% (p=0.001 n=10+5)     873.7µ ± ∞ ¹  -50.83% (p=0.001 n=10+5)
StoreSeries/relabel-32                               1516.5µ ±  7%   658.2µ ± ∞ ¹  -56.60% (p=0.001 n=10+5)     662.5µ ± ∞ ¹  -56.32% (p=0.001 n=10+5)
StoreSeries/externalLabels+relabel-32                 2.122m ±  6%   1.109m ± ∞ ¹  -47.77% (p=0.001 n=10+5)     1.149m ± ∞ ¹  -45.87% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/plain-32                    1743.9µ ± 10%   367.6µ ± ∞ ¹  -78.92% (p=0.001 n=10+5)     374.6µ ± ∞ ¹  -78.52% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/externalLabels-32            3.545m ±  6%   1.600m ± ∞ ¹  -54.86% (p=0.001 n=10+5)     1.637m ± ∞ ¹  -53.82% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/relabel-32                   2.827m ± 14%   1.265m ± ∞ ¹  -55.26% (p=0.001 n=10+5)     1.284m ± ∞ ¹  -54.58% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/externalLabels+relabel-32    4.097m ±  8%   2.161m ± ∞ ¹  -47.25% (p=0.001 n=10+5)     2.246m ± ∞ ¹  -45.18% (p=0.001 n=10+5)
BuildWriteRequest/2_instances-32                      19.02µ ±  3%
BuildWriteRequest/10_instances-32                     89.51µ ±  5%
BuildWriteRequest/1k_instances-32                     915.0µ ±  1%
geomean                                               1.063m         815.7µ        -61.04%                  ²   836.6µ        -60.04%                  ²
¹ need >= 6 samples for confidence interval at level 0.95
² benchmark set differs from baseline; geomeans may not be comparable

                                                   │      base      │                 no-intern                 │               split-intern                │
                                                   │      B/op      │     B/op       vs base                    │     B/op       vs base                    │
SampleSend-32                                         682.6Ki ±  9%
StoreSeries/plain-32                                  415.1Ki ±  0%   271.4Ki ± ∞ ¹  -34.61% (p=0.001 n=10+5)     271.4Ki ± ∞ ¹  -34.61% (p=0.001 n=10+5)
StoreSeries/externalLabels-32                        1040.7Ki ±  0%   897.0Ki ± ∞ ¹  -13.81% (p=0.001 n=10+5)     897.0Ki ± ∞ ¹  -13.81% (p=0.001 n=10+5)
StoreSeries/relabel-32                                916.1Ki ±  0%   772.5Ki ± ∞ ¹  -15.68% (p=0.001 n=10+5)     772.5Ki ± ∞ ¹  -15.68% (p=0.001 n=10+5)
StoreSeries/externalLabels+relabel-32                1010.1Ki ±  0%   866.4Ki ± ∞ ¹  -14.22% (p=0.001 n=10+5)     866.4Ki ± ∞ ¹  -14.22% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/plain-32                     809.7Ki ±  0%   522.3Ki ± ∞ ¹  -35.49% (p=0.001 n=10+5)     522.4Ki ± ∞ ¹  -35.49% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/externalLabels-32            2.013Mi ±  0%   1.732Mi ± ∞ ¹  -13.95% (p=0.001 n=10+5)     1.732Mi ± ∞ ¹  -13.94% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/relabel-32                   1.769Mi ±  0%   1.489Mi ± ∞ ¹  -15.85% (p=0.001 n=10+5)     1.489Mi ± ∞ ¹  -15.85% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/externalLabels+relabel-32    1.953Mi ±  0%   1.672Mi ± ∞ ¹  -14.37% (p=0.001 n=10+5)     1.672Mi ± ∞ ¹  -14.37% (p=0.001 n=10+5)
BuildWriteRequest/2_instances-32                        80.00 ±  0%
BuildWriteRequest/10_instances-32                       80.00 ±  0%
BuildWriteRequest/1k_instances-32                       80.00 ± 15%
geomean                                               98.02Ki         890.1Ki        -20.28%                  ²   890.1Ki        -20.28%                  ²
¹ need >= 6 samples for confidence interval at level 0.95
² benchmark set differs from baseline; geomeans may not be comparable

                                                   │     base     │                no-intern                 │               split-intern               │
                                                   │  allocs/op   │  allocs/op    vs base                    │  allocs/op    vs base                    │
SampleSend-32                                          936.0 ± 9%
StoreSeries/plain-32                                  1666.0 ± 0%    594.0 ± ∞ ¹  -64.35% (n=10+5)              594.0 ± ∞ ¹  -64.35% (n=10+5)
StoreSeries/externalLabels-32                         2.670k ± 0%   1.595k ± ∞ ¹  -40.26% (n=10+5)             1.594k ± ∞ ¹  -40.30% (n=10+5)
StoreSeries/relabel-32                                5.665k ± 0%   4.595k ± ∞ ¹  -18.89% (n=10+5)             4.595k ± ∞ ¹  -18.89% (n=10+5)
StoreSeries/externalLabels+relabel-32                 5.669k ± 0%   4.595k ± ∞ ¹  -18.95% (n=10+5)             4.595k ± ∞ ¹  -18.95% (n=10+5)
StoreSeries_TwoEndpoints/plain-32                     2870.0 ± 0%    727.0 ± ∞ ¹  -74.67% (n=10+5)              727.0 ± ∞ ¹  -74.67% (n=10+5)
StoreSeries_TwoEndpoints/externalLabels-32            4.878k ± 0%   2.728k ± ∞ ¹  -44.08% (p=0.001 n=10+5)     2.727k ± ∞ ¹  -44.10% (p=0.001 n=10+5)
StoreSeries_TwoEndpoints/relabel-32                  10.868k ± 0%   8.729k ± ∞ ¹  -19.68% (n=10+5)             8.729k ± ∞ ¹  -19.68% (n=10+5)
StoreSeries_TwoEndpoints/externalLabels+relabel-32   10.876k ± 0%   8.729k ± ∞ ¹  -19.74% (n=10+5)             8.729k ± ∞ ¹  -19.74% (n=10+5)
BuildWriteRequest/2_instances-32                       1.000 ± 0%
BuildWriteRequest/10_instances-32                      1.000 ± 0%
BuildWriteRequest/1k_instances-32                      1.000 ± 0%
geomean                                                495.9        2.723k        -42.01%                  ²   2.723k        -42.02%                  ²
¹ need >= 6 samples for confidence interval at level 0.95
² benchmark set differs from baseline; geomeans may not be comparable

                                  │       base        │
                                  │ compressedSize/op │
BuildWriteRequest/2_instances-32          1.933k ± 0%
BuildWriteRequest/10_instances-32         8.156k ± 0%
BuildWriteRequest/1k_instances-32         78.22k ± 0%
geomean                                   10.72k
```

### CODE REVIEW

**Comment by cstyan** [comment]
updated benchmark, removed the option to intern if there's more than one queue since it's simpler that way/it would be hard to do correctly:
```

**Comment by cstyan** [comment]
`

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -702,8 +699,8 @@ outer:
 			continue
 		}
 		t.seriesMtx.Lock()
-		lbls, ok := t.seriesLabels[s.Ref]
-		if !ok {
+		lbls := t.seriesLabels[s.Ref]
+		if lbls.Len() == 0 {
```
Comment (cstyan) [inline storage/remote/queue_manager.go:703]: this is no longer needed to change, I will change it back to checking `ok` in the morning

**Comment by bboreham** [comment]
Are you running the benchmarks with `-tags stringlabels` which is the default for Prometheus release builds?

**Review by bwplotka — COMMENTED** [review]
I am supportive, but let's benchmark with new labels mechanism as @bboreham suggested.

**Comment by cstyan** [comment]
> Are you running the benchmarks with `-tags stringlabels` which is the default for Prometheus release builds?

I was not, here are those results: 
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/storage/remote
cpu: AMD Ryzen 9 5950X 16-Core Processor
                                                   │    base     │              no-intern              │
                                                   │   sec/op    │   sec/op     vs base                │
SampleSend-32                                        3.701m ± 5%   4.253m ± 6%  +14.89% (p=0.000 n=10)
StoreSeries/plain-32                                 274.1µ ± 1%   265.6µ ± 1%   -3.08% (p=0.000 n=10)
StoreSeries/externalLabels-32                        755.6µ ± 1%   748.3µ ± 3%        ~ (p=0.280 n=10)
StoreSeries/relabel-32                               962.2µ ± 1%   982.9µ ± 7%   +2.16% (p=0.000 n=10)
StoreSeries/externalLabels+relabel-32                1.098m ± 1%   1.087m ± 1%   -0.97% (p=0.002 n=10)
StoreSeries_TwoEndpoints/plain-32                    510.7µ ± 4%   495.6µ ± 1%   -2.94% (p=0.001 n=10)
StoreSeries_TwoEndpoints/externalLabels-32           1.478m ± 1%   1.469m ± 2%        ~ (p=0.123 n=10)
StoreSeries_TwoEndpoints/relabel-32                  1.877m ± 0%   1.885m ± 3%        ~ (p=0.143 n=10)
StoreSeries_TwoEndpoints/externalLabels+relabel-32   2.136m ± 1%   2.175m ± 7%   +1.86% (p=0.005 n=10)
BuildWriteRequest/2_instances-32                     19.33µ ± 2%   19.44µ ± 3%        ~ (p=0.218 n=10)
BuildWriteRequest/10_instances-32                    92.02µ ± 1%   91.78µ ± 1%        ~ (p=0.353 n=10)
BuildWriteRequest/1k_instances-32                    937.2µ ± 3%   941.4µ ± 1%        ~ (p=0.739 n=10)
geomean                                              632.3µ        637.8µ        +0.87%

                                                   │     base      │                no-intern                 │
                                                   │     B/op      │      B/op       vs base                  │
SampleSend-32                                         675.8Ki ± 3%    751.2Ki ± 12%  +11.16% (p=0.000 n=10)
StoreSeries/plain-32                                  227.9Ki ± 0%    227.8Ki ±  0%   -0.05% (p=0.000 n=10)
StoreSeries/externalLabels-32                         978.7Ki ± 0%    978.6Ki ±  0%   -0.01% (p=0.000 n=10)
StoreSeries/relabel-32                               1011.0Ki ± 0%   1010.8Ki ±  0%   -0.01% (p=0.000 n=10)
StoreSeries/externalLabels+relabel-32                1011.0Ki ± 0%   1010.9Ki ±  0%   -0.01% (p=0.000 n=10)
StoreSeries_TwoEndpoints/plain-32                     435.3Ki ± 0%    435.1Ki ±  0%   -0.05% (p=0.000 n=10)
StoreSeries_TwoEndpoints/externalLabels-32            1.892Mi ± 0%    1.891Mi ±  0%   -0.01% (p=0.000 n=10)
StoreSeries_TwoEndpoints/relabel-32                   1.955Mi ± 0%    1.954Mi ±  0%   -0.01% (p=0.000 n=10)
StoreSeries_TwoEndpoints/externalLabels+relabel-32    1.955Mi ± 0%    1.954Mi ±  0%   -0.01% (p=0.000 n=10)
BuildWriteRequest/2_instances-32                        80.00 ± 0%      80.00 ±  0%        ~ (p=1.000 n=10) ¹
BuildWriteRequest/10_instances-32                       80.00 ± 0%      80.00 ±  0%        ~ (p=1.000 n=10) ¹
BuildWriteRequest/1k_instances-32                       80.00 ± 0%      80.00 ±  0%        ~ (p=1.000 n=10) ¹
geomean                                               89.03Ki         89.81Ki         +0.87%
¹ all samples are equal

                                                   │    base     │               no-intern                │
                                                   │  allocs/op  │  allocs/op    vs base                  │
SampleSend-32                                         928.5 ± 3%   1035.0 ± 13%  +11.47% (p=0.000 n=10)
StoreSeries/plain-32                                  603.0 ± 0%    601.0 ±  0%   -0.33% (p=0.000 n=10)
StoreSeries/externalLabels-32                        1.604k ± 0%   1.602k ±  0%   -0.12% (p=0.000 n=10)
StoreSeries/relabel-32                               4.605k ± 0%   4.603k ±  0%   -0.04% (p=0.000 n=10)
StoreSeries/externalLabels+relabel-32                4.605k ± 0%   4.603k ±  0%   -0.04% (p=0.000 n=10)
StoreSeries_TwoEndpoints/plain-32                     745.0 ± 0%    741.0 ±  0%   -0.54% (p=0.000 n=10)
StoreSeries_TwoEndpoints/externalLabels-32           2.746k ± 0%   2.742k ±  0%   -0.15% (p=0.000 n=10)
StoreSeries_TwoEndpoints/relabel-32                  8.748k ± 0%   8.745k ±  0%   -0.03% (p=0.000 n=10)
StoreSeries_TwoEndpoints/externalLabels+relabel-32   8.748k ± 0%   8.745k ±  0%   -0.03% (p=0.000 n=10)
BuildWriteRequest/2_instances-32                      1.000 ± 0%    1.000 ±  0%        ~ (p=1.000 n=10) ¹
BuildWriteRequest/10_instances-32                     1.000 ± 0%    1.000 ±  0%        ~ (p=1.000 n=10) ¹
BuildWriteRequest/1k_instances-32                     1.000 ± 0%    1.000 ±  0%        ~ (p=1.000 n=10) ¹
geomean                                               346.4         349.1         +0.80%
¹ all samples are equal

                                  │       base        │                 no-intern                  │
                                  │ compressedSize/op │ compressedSize/op  vs base                 │
BuildWriteRequest/2_instances-32          1.933k ± 0%         1.933k ± 0%       ~ (p=1.000 n=10) ¹
BuildWriteRequest/10_instances-32         8.156k ± 0%         8.156k ± 0%       ~ (p=1.000 n=10) ¹
BuildWriteRequest/1k_instances-32         78.22k ± 0%         78.22k ± 0%       ~ (p=1.000 n=10) ¹
geomean                                   10.72k              10.72k       +0.00%
```

The results with `-tags dedupelabels` are actually pretty bad, but afaict we don't ever build with that tag currently?

So I gues we could change this to allow optional creation/usage of the interning functionality so that projects who are reusing remote write code can benefit from not using it, or I can try to get loki to build with `-tags stringlabels`.

**Comment by jmichalek132** [comment]
Any decisions on yet, as this will have an impact on implementing rw2 support in otel collector remote write exporter. https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33661

**Comment by cstyan** [comment]
@jmichalek132 sorry, I think maybe you've misunderstood. Interning of strings within the actual Remote Write proto is still part of 2.0 RW spec and is required.

This change is an internal implementation detail of Prometheus' own remote write implementation, in which we use a map to cache the labels seen from Prometheus' WAL series records. We used interning in addition to that cache since we used to cache each label name and value individually and then recreate the labelset when collecting all the necessary labels from the cache to build the proto write request.

**Comment by cstyan** [comment]
This change didn't have a large measurable impact where I was testing it, but the profiles do show a similar 30% reduction in memory in-use for a system that is not built with `stringlabels`. Baseline is with these changes + choosing not to intern the labels, comparison is with the interning.
2024-09-25-165824_2469x728_scrot

We could merge this, or alternatively just remove the interning code altogether.

**Comment by bboreham** [comment]
Hi, would it be correct to title this as "allow optional disabling of interning within remote-write" ? 
I got confused initially, thinking it might be adding more interning.

I'm ok to merge the option, or to remove the code altogether.  Prometheus doesn't use it and I can't think of any downstream project that definitely needs it.

**Comment by bboreham** [comment]
PR to delete the code at #15456 - I found the map lookup showed up in an analysis of why remote-write was lagging behind.

**Comment by cstyan** [comment]
closing in favour of Bryans pr :point_up:

**Pull request closed without merging**